jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -331,34 +333,66 @@ object XGBoost extends Serializable {
         " an instance of Long.")
     }
 
-    val tracker = startTracker(nWorkers, trackerConf)
-    try {
-      val sc = trainingData.sparkContext
-      val parallelismTracker = new SparkParallelismTracker(sc, timeoutRequestWorkers, nWorkers)
-      val overriddenParams = overrideParamsAccordingToTaskCPUs(params, trainingData.sparkContext)
-      val boostersAndMetrics = buildDistributedBoosters(trainingData, overriddenParams,
-        tracker.getWorkerEnvs, nWorkers, round, obj, eval, useExternalMemory, missing)
-      val sparkJobThread = new Thread() {
-        override def run() {
-          // force the job
-          boostersAndMetrics.foreachPartition(() => _)
-        }
-      }
-      sparkJobThread.setUncaughtExceptionHandler(tracker)
-      sparkJobThread.start()
-      val isClsTask = isClassificationTask(params)
-      val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
-      logger.info(s"Rabit returns with exit code $trackerReturnVal")
-      val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
-        sparkJobThread, isClsTask)
-      if (isClsTask){
-        model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
-          params.getOrElse("num_class", "2").toString.toInt
+    val hdfsTmpPath: String = params.get("hdfs_tmp_path") match {
+      case None => null
+      case Some(path: String) => if (path.nonEmpty) {
+        path
+      } else {
+        null
       }
-      model
-    } finally {
-      tracker.stop()
+      case _ => throw new IllegalArgumentException("parameter \"hdfs_tmp_path\" must be" +
+        " an instance of String.")
+    }
+
+    val savingFreq: Int = params.get("saving_frequency") match {
+      case None => 0
+      case Some(freq: Int) => freq
+      case _ => throw new IllegalArgumentException("parameter \"saving_frequency\" must be" +
+        " an instance of Int.")
     }
+
+    val savingRounds: Seq[Int] = if (hdfsTmpPath == null || savingFreq <= 0){
```

Comment:
we should not swallow the user's input error here, i.e. we should deal with hdfsTmpPath == null as a wrong param in the helper function above 

and personally, I think the following way to regulate these parameters is more natural to the user

1. if the user sets a > 0 value for savingFreq, that means they want to the checkpoint 
2. if the sets a > 0 value for savingFreq, they must specify hdfsTmpPath, otherwise, we should stop training due to the missing parameter

Reply:
Done.